mypy/checker.py

```diff
@@ -3577,9 +3578,14 @@ def check_subtype(self, subtype: Type, supertype: Type, context: Context,
                     extra_info.append(supertype_label + ' ' + supertype_str)
                 note_msg = make_inferred_type_note(context, subtype,
                                                    supertype, supertype_str)
+                if msg == message_registry.INCOMPATIBLE_TYPES_IN_ASSIGNMENT:
+                    notes = append_invariance_notes([], subtype,  supertype)
```

Comment:
You need to do this _only_ if `isinstance(subtype, Instance)` and `isinstance(supertype, Instance)`, see the signature of `append_invariance_notes`. By the way this error would be caught statically by mypy during self-check. To run it use `python runtests.py self`.

Reply:
Please note that you didn't address this issue, which is probably most important one since it causes all your test failures.

Reply:
I'm fixing it. I'm trying to have a deeper understanding so I can understand why and not just. I want to be able to under better the project and contribute more often
.

Reply:
Now I have just one test failing called `testCoberturaParser`

```
(mypy) ➜  mypy git:(better-error-message-for-assignments-incompatible-due-to-invariance) ✗ pytest --ff
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 3.7.3, pytest-4.4.1, py-1.8.0, pluggy-0.10.0
rootdir: /Users/felipe/projects/python/mypy, inifile: pytest.ini, testpaths: mypy/test
plugins: forked-1.0.2, xdist-1.28.0, cov-2.7.1
gw0 [7577] / gw1 [7577] / gw2 [7577] / gw3 [7577]
...........................s........................................................................................................................................................................ [  2%]
.................................................................................................................................................................................................... [  5%]
.................................................................................................................................................................................................... [  7%]
.......................................................................................................s............................................................................................ [ 10%]
........................................s.....................................................................s..................................................................s.................. [ 12%]
.................................................................................................................................................................................................... [ 15%]
.....................................................................s....................................................................................................s......................... [ 18%]
...............................................................................................................................................................................s.................... [ 20%]
..........................................s.................................................s........................................................................................................ [ 23%]
.............................................................s...............................ss.s.........................................s......................................................... [ 25%]
............................................................s....................................................................................................................................... [ 28%]
..................................................................................................................................................................................................... [ 30%]
..................................................................................................................................................................................................... [ 33%]
.................................................................................................................................................................................................... [ 36%]
.................................................................................................................................................................................................... [ 38%]
.................................................................................................................................................................................................... [ 41%]
.................................................................................................................................................................................................s.. [ 43%]
...........s........................................................................................................................................................................................ [ 46%]
...........................s...........................................................s........................................................................s................................... [ 48%]
...................s.........................................................................................................ss..................................................................... [ 51%]
..................................................................................................................s................................................................................. [ 54%]
...............................................................................................................................................................................s.................... [ 56%]
.........................................................................s.....................s......................s............................................................................. [ 59%]
...................sss.sssss................................................................................................s................................................s...................... [ 61%]
......................................................................s.................s...................s.......sss.......s....s.......s.........s...................s...sss.s................s. [ 64%]
.....s............................s........s...........s........................s.ss.................sss......sssssssss.....................sss..s....ss......ss.s.s........s.s.....s....s...ss..... [ 66%]
s.s.....ss..................................s................s......s..............s.................ssss.....sss....sssssss......s...s...s.................................s....................... [ 69%]
...................................sss.ss.....................................s.s................................................................................................................... [ 72%]
............s........s..............................s................................................................sss..........................................................s.....s...ssssssss [ 74%]
s.sssssss.s.ssssssssss.ss......................................................................................................s..s......................................s..s...s.ss.s.............. [ 77%]
....s.s.....s.....s..s.sss.........s...ssss.s.ss.....ss............................................................................................................................................. [ 79%]
.................................................................................................................................................................................................... [ 82%]
.................................................................................................................................................................................................... [ 84%]
.................................................................................................................................................................................................... [ 87%]
.................................................................................................................................................................................................... [ 89%]
..................................................................................................................................................................................s................. [ 92%]
..............................................................................s....sss..........................s.....................................s............................................. [ 95%]
...............................s...........................................................F........................................................................................................ [ 97%]
............................................................................s.................................................
============================================= FAILURES ==============================================
________________________________________ testCoberturaParser ________________________________________
[gw2] darwin -- Python 3.7.3 /Users/felipe/.local/share/virtualenvs/mypy-WvJTD_Lt/bin/python3.7
data: /Users/felipe/projects/python/mypy/test-data/unit/reports.test:17:
/Users/felipe/projects/python/mypy/mypy/test/testcmdline.py:36: in run_case
    test_python_cmdline(testcase, step)
/Users/felipe/projects/python/mypy/mypy/test/testcmdline.py:100: in test_python_cmdline
    path, ' on step %d' % step if testcase.output2 else ''))
/Users/felipe/projects/python/mypy/mypy/test/helpers.py:115: in assert_string_arrays_equal
    raise AssertionError(msg)
E   AssertionError: Output file tmp/build/cobertura.xml did not match its expected output
--------------------------------------- Captured stderr call ----------------------------------------
Expected:
  <coverage timestamp="$TIMESTAMP" version="$VERSION" line-rate="0.8000" branch-rate="0"> (diff)
    <sources>
      <source>$PWD</source>
    </sources>
    <packages>
  ...
Actual:
  <coverage timestamp="$TIMESTAMP" version="$VERSION.62a6c672c8545399e7f3dec328b5a910e1dc5f26" line-rate="0.8000" branch-rate="0"> (diff)
    <sources>
      <source>$PWD</source>
    </sources>
    <packages>
  ...

Alignment of first line difference:
  E: ...P" version="$VERSION" line-rate="0.8000" branch-rate="0">...
  A: ...P" version="$VERSION.62a6c672c8545399e7f3dec328b5a910e1dc5f26" line-r...
                            ^
========================================= warnings summary ==========================================
main:2
  main:2: DeprecationWarning: invalid escape sequence \u

main:1
  main:1: DeprecationWarning: <> not supported in 3.x; use !=

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================= 1 failed, 7381 passed, 195 skipped, 2 warnings in 617.66 seconds ==================
```

Reply:
This is not a real failure (for example you don't see it in CI), this is just because you made a commit while test was running. The only failure is currently failing flake8 linter (run `python runtests.py lint` to run locally).

Reply:
Thank you. I fixed the flake8 error.